300-CNAB-security.md

```diff
@@ -0,0 +1,163 @@
+
+This specification does not specifically deal with network-layer (Layer 7) security, but assumes that implementations will apply correct security constraints for all network traffic. TLS/SSL is likely the method by which such traffic should be secured.
+
+### Overview
+
+There are three major aspects of securely working with registries:
+
+1. Pulling bundles _from_ a registry _to_ a client
+2. Pushing bundles _from_ a client _to_ a registry
+3. Managing keys between a client and a registry
+
+### Pulling and Verifying Bundles
+
+When a client pulls a bundle from a repository, the client MUST verify the package, where verification consists of the following:
+
+- Fetch (from a TUF endpoint) the signature object for the given bundle
+- Fetch the bundle descriptor (optionally by fetching a thick bundle and unpacking the descriptor)
+- Verify the requisite signatures for a package, as described in [the TUF specification](https://github.com/theupdateframework/specification/blob/master/tuf-spec.md#5-detailed-workflows).
+- Verify that each Invocation Image in the bundle descriptor (a) has a digest, and (b) the digest matches the referenced image
+- Verify that each Image referenced in the bundle descriptor (a) has a digest, and (b) the digest matches the referenced image
+
+A client MUST be able to verify SHA-256 and SHA-512 digests. But this specification accords with the CNAB Core specification in that it does not dictate what are the the contents of the Invocation Images or regular Images. Examples in this document often refer to OCI images as the content of these image fields, but other formats are acceptable.
+
+If any of these steps fail to obtain, the client MUST NOT perform an action on the bundle. Clients SHOULD produce an error.
+
+### Signing and Pushing Bundles
+
+When a client pushes a bundle to a repository, the client MUST prepare the security data on the package, which consists of the following:
+
+- Generate the digest (SHA-256 or SHA-512) of each invocation image, and write this information into the bundle descriptor
+- Generate the digest of each image, and write this information into the bundle descriptor
+- Optionally, push the images to their appropriate remote location (thin bundles)
+- Sign the requisite portion of the bundle
+	- For thin bundles, this is the bundle descriptor alone
+	- For thick bundles, this is the full bundle, in its representation that it will be submitted to the bundle registry
+- Send the signature data to the TUF server, per the details of that server's implementation. (The TUF specification does not specify the process for uploading signatures. However, Notary provides one API for doing this.) TODO: Do we want to limit this to the Notary-prescribed protocol for ease of developing compatible clients?
```

Comment:
👍 did you mean to leave this TODO in here? I'm good with that. Just wanted to make sure it's on purpose.

Reply:
Yeah, until I figure out some of the details.